### PR TITLE
Handle web socket downgrade when HTTP/2 not enabled

### PR DESCRIPTION
**Pull request by greenEkatherine**
During manual testing of https://github.com/dotnet/runtime/pull/73762 @BrennanConroy and me detected that in the case when HTTP/2 connection is not supported downgrade for web sockets is not working as expected.

In that case `ClientWebSocket` didn't get exception data indicating that downgrade is possible, so I added this data and tests for it.

However, it doesn't handle plain text scenario because HttpClient in that case silently create Http2Connection with no chance to detect the real version before we try to establish extended connect. If downgrade allowed by policy HttpClient force 1.1 without giving 2.0 a try -- maybe ClientWebSocket should do the same.

### CODE REVIEW

File: src/libraries/System.Net.WebSockets.Client/tests/ConnectTest.Http2.cs [hunk 1]
```diff
@@ -96,6 +96,57 @@ public async Task ConnectAsync_VersionNotSupported_WithSsl_Throws()
...
+                if (useHandler)
+                {
+                    var handler = new SocketsHttpHandler();
+                    await cws.ConnectAsync(server, new HttpMessageInvoker(handler), cts.Token);
+                }
+                else
+                {
+                    await cws.ConnectAsync(server, cts.Token);
+                }
+            }
```
Comment (CarnaViire) [inline src/libraries/System.Net.WebSockets.Client/tests/ConnectTest.Http2.cs:146]: Let's also assert that cws in correct state after connecting

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs [hunk 2]
```diff
@@ -88,7 +88,12 @@ public static async ValueTask<SslStream> EstablishSslConnectionAsync(SslClientAu
                     throw CancellationHelper.CreateOperationCanceledException(e, cancellationToken);
                 }
 
-                throw new HttpRequestException(SR.net_http_ssl_connection_failed, e);
+                HttpRequestException ex = new HttpRequestException(SR.net_http_ssl_connection_failed, e);
+                if (request.IsExtendedConnectRequest)
+                {
+                    ex.Data["HTTP2_ENABLED"] = false;
```
Comment (CarnaViire) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:94]: It seems a bit strange given that the code seems to be about SSL in general, not HTTP/2 specific; and it is also possible to have plaintext HTTP/2.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 3]
```diff
@@ -423,7 +423,13 @@ private static void ThrowGetVersionException(HttpRequestMessage request, int des
         {
             Debug.Assert(desiredVersion == 2 || desiredVersion == 3);
 
-            throw new HttpRequestException(SR.Format(SR.net_http_requested_version_cannot_establish, request.Version, request.VersionPolicy, desiredVersion), inner);
+            HttpRequestException ex = new HttpRequestException(SR.Format(SR.net_http_requested_version_cannot_establish, request.Version, request.VersionPolicy, desiredVersion), inner);
+            if (request.IsExtendedConnectRequest)
```
Comment (CarnaViire) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:427]:
```suggestion
            if (request.IsExtendedConnectRequest && desiredVersion == 2)
```

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 4]
```diff
@@ -1119,12 +1125,7 @@ public async ValueTask<HttpResponseMessage> SendWithVersionDetectionAndRetryAsyn
                     // Throw if fallback is not allowed by the version policy.
                     if (request.VersionPolicy != HttpVersionPolicy.RequestVersionOrLower)
                     {
-                        HttpRequestException exception = new HttpRequestException(SR.Format(SR.net_http_requested_version_server_refused, request.Version, request.VersionPolicy), e);
-                        if (request.IsWebSocketH2Request())
-                        {
-                            exception.Data["HTTP2_ENABLED"] = false;
-                        }
-                        throw exception;
+                        throw new HttpRequestException(SR.Format(SR.net_http_requested_version_server_refused, request.Version, request.VersionPolicy), e);
```
Comment (CarnaViire) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1128]: Why did you remove the setting here?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs (hunk 2 above)
Comment (CarnaViire) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:94]: We might want to have another name for Data here; I would also expect if SSL is not supported for the server, but we want a secure websocket connection ("wss") -- it doesn't make sense to try 1.1 after that because it would fail again? Unless I miss something...

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs (hunk 4 above)
Comment (greenEkatherine) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1128]: Because we never go there during downgrade from HTTP/2. The only place where this type of retry is set is in Http3Stream. I might add this type of retry on the exceptions where I now add exception data but that would be one more hop and not transparent logic.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs (hunk 2 above)
Comment (greenEkatherine) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:94]: `sslOptions` is HTTP/2 specific in this method, we try to get it for exact HTTP/2 without knowledge that downgrade is possible -- it is out of scope for HttpClient

Comment (greenEkatherine) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:94]: Why it doesn't make sense to try 1.1? Secure websocket connection is supported for it

Comment (CarnaViire) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:94]: I did some code investigation and now I understand, the SSL connection was trying to be established for strictly ALPN="h2", so ALPN="http/1.1" was not allowed by client and that's why SSL connection failed

Comment (CarnaViire) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:94]: Maybe worth adding a comment about it

**Review by CarnaViire — APPROVED** [review]
LGTM, thanks!

**Comment by Tratcher** [comment]
> If downgrade allowed by policy HttpClient force 1.1 without giving 2.0 a try -- maybe ClientWebSocket should do the same.

Yes, ClientWebSocket should follow the same rules as HttpClient here.

**Comment by MihaZupan** [comment]
I agree we should match `SocketsHttpHandler`
https://github.com/dotnet/runtime/blob/3db2ebfcfa2b5c7074a7e513a56a39f07f6ce6c0/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1037-L1039

If the scheme is not secure and the policy allows 1.1, we should not attempt 2.0.

**Comment by greenEkatherine** [comment]
I added condition to force to 1.1 non-secure web socket requests in ClientWebSocket @MihaZupan @Tratcher

File: src/libraries/System.Net.WebSockets.Client/tests/ConnectTest.Http2.cs [hunk 5]
```diff
@@ -96,6 +96,59 @@ public async Task ConnectAsync_VersionNotSupported_WithSsl_Throws()
...
+                var ex = await Assert.ThrowsAnyAsync<WebSocketException>(() => t);
+                Assert.IsType<HttpRequestException>(ex.InnerException);
+                Assert.True(ex.InnerException.Data.Contains("HTTP2_ENABLED"));
+                Assert.Equal(WebSocketState.Closed, cws.State);
+            }
+        }
+
+        [OuterLoop("Uses external servers", typeof(PlatformDetection), nameof(PlatformDetection.LocalEchoServerIsNotAvailable))]
+        [Theory]
+        [MemberData(nameof(EchoServersAndBoolean))]
```
Comment (CarnaViire) [inline src/libraries/System.Net.WebSockets.Client/tests/ConnectTest.Http2.cs:129]: Do EchoServersAndBoolean include secure server as well?

Comment (greenEkatherine) [reply src/libraries/System.Net.WebSockets.Client/tests/ConnectTest.Http2.cs:129]: Yes, it includes both

**Review by CarnaViire — APPROVED** [review]

**Comment by MihaZupan** [comment]
Does HTTP/2 over cleartext work with WebSockets right now?

SocketsHttpHandler requires that you use set both Version=2.0 and Policy=Exact for that.
From what I can see, we set only the version in WebSockets: https://github.com/dotnet/runtime/blob/feb0e27659432640b3858e7330d75b124c198499/src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs#L73

**Comment by greenEkatherine** [comment]
> Does HTTP/2 over cleartext work with WebSockets right now?
> 
> SocketsHttpHandler requires that you use set both Version=2.0 and Policy=Exact for that. From what I can see, we set only the version in WebSockets:
> 
> https://github.com/dotnet/runtime/blob/feb0e27659432640b3858e7330d75b124c198499/src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs#L73

VersionPolicy is set up later in the `AddWebSocketHeaders` https://github.com/dotnet/runtime/blob/feb0e27659432640b3858e7330d75b124c198499/src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs#L370

**Comment by greenEkatherine** [comment]
Failures in the outerloop builds are unrelated
